CompilerSource/parsing/pretty_printer.cpp

```diff
@@ -0,0 +1,733 @@
+/** Copyright (C) 2024 Fares Atef
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include <JDI/src/System/builtins.h>
+#include "ast.h"
+
+using namespace enigma::parsing;
+
+bool AST::Visitor::VisitIdentifierAccess(IdentifierAccess &node) {
+  print(std::string(node.name.content));
+  return true;
+}
+
+bool AST::Visitor::VisitLiteral(Literal &node) {
+  std::string value = std::get<std::string>(node.value.value);
+  if (value == "'\n'") {
```

Comment:
This won't pass much more than your test cases.

EDL does some backflips to reflect how GML does things. We should have a setting somewhere for which language (C++ or GML) literals are inherited from. In this case, I don't think the onus should fall on the pretty printer to *decode* the string, but to encode it. So  basically, literals should know what type they have (char, string, int, float, etc) and what the final value is. So during *parse*, we should be checking if the input string was `"` or `'` (single or double-quote), and if single ***and*** literals are inherited from C++, interpret as a char literal (and warn if >1 char in the end). Similarly, that parse method should handle interpreting escapes. `\n` doesn't mean newline in GML; `#` means newline. This is dumb IMO, so it's configurable in ENIGMA.

So the parser (particularly the lexer) is already checking those settings (where literals are inherited from, where escapes are inherited from). It already uses those settings to determine whether `'a'` is a string literal (`TT_STRINGLIT`) or char literal (`TT_CHARLIT`). What I don't think it's doing is computing the *raw* value of these literals in the parser itself. In those routines, it should handle decoding those values:
```cpp
if (value.length() < 2 || value[0] != value[value.length() - 1]) return error(Internal error: Invalid string literal);
std::string str_value;
str_value.reserve(value.length() - 2);
if (options.use_escapes) {
  for (size_t i = 1; i < value.length() - 1; ++i) {
    if (value[i] == '\\') {
      if (++i >= value.length())
        return error(Internal error: lexer stopped parsing string literal early);
      switch (value[++i]) {
        case '\\': str_value += '\\'; break;
        case 'n': str_value += '\n'; break;
        // ...
        default: return error(Unkown escape); // or warn instead and insert it raw; do what your heart tells you
      }
    } else {
      str_value += value[i];
    }
  }
} else {
  for (size_t i = 1; i < value.length() - 1; ++i) {
    str_value += value[i] == '#' ? '\n' : value[i];
  }
}
```

Then in this routine, you encode the raw string:
```cpp
print(type == TT_CHARLIT ? '\'' : '\"');
for (char c : str_value) {
  if (c == '\\') {
    print("\\\\");
  } else if (c >= ' ' and c <= '~') {
    print(c);
  } else if (c == '\n') {
    print("\\n");
  } else if (c == '\r') {
    print("\\r");
  } else {
    std::ostringstream oss;
    oss << '\\' << std::oct << static_cast<int>(c);
    print(oss.str());
  }
}
print(type == TT_CHARLIT ? '\'' : '\"');
```

(Code is approximate/pseudo-code.)

Reply:
Thanks, Josh, for this detailed explanation. I will handle this in the upcoming PR, along with the previously mentioned item in the to-do list: parse single and double quotes.